Bucketmax write strategy

piotr1212 opened the pull request:
Bucketmax writes the metric with the most datapoints just as max does but uses
a different approach to determine which metric has the most datapoints.

At each write operation 'max' loops once over the list of metrics and picks the metric
with most datapoints.

'bucketmax' stores all metrics in a list where each index represents the number of
datapoints in the specific metrics. For example `buckets[5]` contains a list of all
metrics which have 6 datapoints. At each write operation 'bucketmax' selects the
first metric from the largest bucket.

The time complexity of 'max' is linear, 'bucketmax' is near constant time but runs
on every ingestion where 'max' only runs on every write. There is also some overhead
in resizing the buckets list. I've done some limited testing and 'bucketmax'' CPU
usage was almost half of that of 'max'. The constraint of bucketmax is that the number
of datapoints per metrics must be relatively low compared to the total number of
unique metrics (which I think is the most common use-case in Graphite usage). I don't
know where the tipping point is but if one has for example 10 metrics with each 100000
datapoints I would assume 'bucketmax' would perform worse than 'max'.

deniszh commented on conf/carbon.conf.example:
```diff
@@ -177,6 +177,9 @@ LOG_CACHE_QUEUE_SORTS = False
 #
+# bucketmax (experimental) - As 'max' but uses a different algorithm to determine
+# the metric with the most datapoints. Should perform better than 'max'.
+#
```
Maybe you should mention " number of datapoints per metrics vs total number of
unique metrics" from PR description here too.

piotr1212 replied:
I guess so. Thing is that that statement is an assumption. This whole thing needs to be better tested to see how it behaves in real life and if my assumption is correct.

deniszh replied:
Sure, my idea was to extend that a bit for users, e.g. "Should perform better than 'max' if number of datapoints per metric is relatively low compared to the total number of
unique metrics, which should be common use-case. See https://github.com/graphite-project/carbon/pull/879 for details."
But that's optional ofc.

piotr1212 replied:
added some more info

deniszh commented:
Great idea, I like that, @piotr1212 !
For me it looks like much proper `max` stratege, I mean, for typical use-case.

DanCech commented:
I don't see where the metric is removed from the associated bucket when it is flushed?

I'm also concerned about the impact on memory usage of maintaining the buckets, since you're going to have a second copy of all the metric names in the bucket data structure.

piotr1212 commented:
> I don't see where the metric is removed from the associated bucket when it is flushed?

It is being popped here: https://github.com/graphite-project/carbon/pull/879/files#diff-b95553eddf4a9a25750a9a1d17f2be0cR161

> I'm also concerned about the impact on memory usage of maintaining the buckets, since you're going to have a second copy of all the metric names in the bucket data structure.

Good point, i'll see if I can store pointers instead of the full string somehow.

piotr1212 commented:
It seems that Python does not copy the metric name but just stores a reference to the string object.

See below for a small simulation, the refcount to the metric name increases when I add it to the buckets. The size of the buckets is smaller than the actual size of the string.


```
In [62]: a_metric_name = 'a very long string representing a metric name in graphite, foobar, some more text, even more and more and more and more'            

In [63]: sys.getsizeof(a_metric_name)  
Out[63]: 168

In [64]: sys.getrefcount(a_metric_name)
Out[64]: 2

In [65]: a_metric_cache = dict()       

In [66]: sys.getsizeof(a_metric_cache) 
Out[66]: 248

In [67]: a_metric_cache[a_metric_name] = [1, 2, 3, 4, 5]                       

In [68]: sys.getsizeof(a_metric_cache) 
Out[68]: 248

In [69]: sys.getrefcount(a_metric_name)
Out[69]: 3

In [70]: buckets = list()              

In [71]: sys.getsizeof(buckets)        
Out[71]: 72

In [72]: buckets.append(a_metric_name) 

In [73]: sys.getsizeof(buckets)        
Out[73]: 104

In [74]: sys.getrefcount(a_metric_name)
Out[74]: 4
```

DanCech commented:
That's good to know.  The only other comment I'd have would be around how we're now mingling handling between the strategy and the cache, and that the cache has strategy-specific code in it.

One thought was that the buckets could be defined within the strategy and a method on the strategy called each time we add a new point so that it could do any housekeeping.  That would remove the strategy-specific code from the cache.

piotr1212 commented:
> That's good to know. The only other comment I'd have would be around how we're now mingling handling between the strategy and the cache, and that the cache has strategy-specific code in it.
> 
> One thought was that the buckets could be defined within the strategy and a method on the strategy called each time we add a new point so that it could do any housekeeping. That would remove the strategy-specific code from the cache.



Thank you for the suggestion. Updated, code looks much cleaner now.


> That's good to know. The only other comment I'd have would be around how we're now mingling handling between the strategy and the cache, and that the cache has strategy-specific code in it.
> 
> One thought was that the buckets could be defined within the strategy and a method on the strategy called each time we add a new point so that it could do any housekeeping. That would remove the strategy-specific code from the cache.

Thank you for the suggestion. Updated it, code looks much cleaner now.

DanCech commented on lib/carbon/cache.py:
```diff
@@ -146,6 +149,41 @@ def choose_item(self):
+        while nr_points > len(self.buckets):
+            self.buckets.append(list())
+
+        # Remove existing metrics from it's bucket
```
```suggestion
        # Remove existing metrics from its bucket
```

piotr1212 commented:
squashed the commit

deniszh commented:
Cool! will try to run in on https://github.com/graphite-project/graphite-test-docker this weekend

deniszh commented:
Ok, I made graph below on my MBP 2017 (docker had 4 cores/6GB RAM) using https://github.com/graphite-project/graphite-test-docker
haggar generated -agents=10 -metrics=10000 = 100K metrics
```
MAX_CACHE_SIZE = Inf
MAX_UPDATES_PER_SECOND = 5000
MAX_CREATES_PER_MINUTE = 50
```

`CACHE_WRITE_STRATEGY = max`
<img width="1366" alt="Screenshot 2020-02-08 at 18 49 27" src="https://user-images.githubusercontent.com/1227222/74101216-be86a100-4b37-11ea-8fcc-b3d0c86dc672.png">
<img width="1354" alt="Screenshot 2020-02-08 at 18 49 40" src="https://user-images.githubusercontent.com/1227222/74101218-c47c8200-4b37-11ea-8736-90d9e5c627bc.png">
<img width="1370" alt="Screenshot 2020-02-08 at 18 50 04" src="https://user-images.githubusercontent.com/1227222/74101219-c5adaf00-4b37-11ea-9963-775332dd9208.png">

`CACHE_WRITE_STRATEGY = bucketmax`
<img width="1380" alt="Screenshot 2020-02-08 at 23 35 57" src="https://user-images.githubusercontent.com/1227222/74101224-d52cf800-4b37-11ea-9aea-ca86e39ac129.png">
<img width="1348" alt="Screenshot 2020-02-08 at 23 36 15" src="https://user-images.githubusercontent.com/1227222/74101225-d65e2500-4b37-11ea-9097-3a4feab7b505.png">
<img width="1363" alt="Screenshot 2020-02-08 at 23 36 33" src="https://user-images.githubusercontent.com/1227222/74101226-d6f6bb80-4b37-11ea-9b58-ff25dd45409b.png">

piotr1212 commented:
But CPU is nearly identical, memory usage is higher. That is quite different from what I measured. I'll do some more testing

piotr1212 commented:
50k metrics every 10s, metrics were already existing.

carbon.conf
``` 
MAX_CACHE_SIZE = inf
MAX_UPDATES_PER_SECOND = 500
MAX_CREATES_PER_MINUTE = 1000000
```

load generator is  https://github.com/Civil/graphite_perf_test_go
```
~/go/bin/graphite_perf_test_go -host 127.0.0.1:2003 -connec
tions 50 -interval 10                                                          
2020/02/09 16:45:33 Starting...                                                
2020/02/09 16:45:33 ===New iteration===                  
2020/02/09 16:45:33 Load    :  50 x 1000 = 50000  metrics
2020/02/09 16:45:34 Spent   :  0.043315 seconds          
2020/02/09 16:45:34 Speed   :  1154334.5 metrics/second  
2020/02/09 16:45:34 Sleeping:  9.956684 seconds          
```


first 30m is bucketmax, then 15m max, then bucketmax again. You see a significant lower CPU usage with bucketmax.

![b_max](https://user-images.githubusercontent.com/2911066/74105338-2e5b5280-4b5d-11ea-93ef-b1f2ebd0756f.png)

deniszh commented:
> But CPU is nearly identical, memory usage is higher. That is quite different from what I measured. I'll do some more testing

Well, memory is the same, I just waited more time. But you're right, some owerflowing test has much more sense here.

piotr1212 commented:
Ok, merging this.

Dieterbe commented:
Is there an explanation for the significantly reduced pointsPerUpdate? I thought the behavior would be equivalent.

piotr1212 commented:
With the 'max' strategy CPU goes to 100%. So I assume it does less write
operations which results in larger queues so more points per update. I
probably should have added the 'write_operations' to the graph and do a
test where it does not overload the CPU.




On Thu, 21 May 2020, 20:45 Dieter Plaetinck, <notifications@github.com>
wrote:

> Is there an explanation for the significantly reduced pointsPerUpdate? I
> thought the behavior would be equivalent.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/graphite-project/carbon/pull/879#issuecomment-632277284>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAWGWWSSJ7SFJPOV2CXJ76TRSVZE7ANCNFSM4KPXYTRQ>
> .
>

Dieterbe commented:
Ah that makes sense and explains why pointsPerUpdate grows linearly.  This also means the memory usage in non overload scenario is likely about the same as judging from the start of the 'max's run
